shared/middleware/net_sim_api.rb

```diff
@@ -0,0 +1,111 @@
+require 'sinatra/base'
+require 'cdo/db'
+require 'cdo/rack/request'
+require 'csv'
+
+class NetSimApi < Sinatra::Base
+
+  helpers do
+    [
+        'core.rb',
+        'storage_id.rb',
+        'table.rb',
+    ].each do |file|
+      load(CDO.dir('shared', 'middleware', 'helpers', file))
+    end
+  end
+
+  TableType = CDO.use_dynamo_tables ? DynamoTable : Table
+
+  # Environment-dependent channel ID for NetSim.
+  # Will no longer be needed when we detach from the channels API entirely.
+  def netsim_channel_id
+    rack_env?(:development) ? 'JGW2rHUp_UCMW_fQmRf6iQ==' : 'HQJ8GCCMGP7Yh8MrtDusIA=='
+  end
+
+  def channels_table_name(shard_id, table_name)
+    "#{shard_id}_#{table_name}"
+  end
+
+  #
+  # GET /v3/netsim/<shard-id>/<table-name>
+  #
+  # Returns all of the rows in the table.
+  #
+  get %r{/v3/netsim/([^/]+)/(\w+)$} do |shard_id, table_name|
+    dont_cache
+    content_type :json
+    TableType.new(netsim_channel_id, storage_id('shared'), channels_table_name(shard_id, table_name)).to_a.to_json
+  end
+
+  #
+  # GET /v3/netsim/<shard-id>/<table-name>/<row-id>
+  #
+  # Returns a single row by id.
+  #
+  get %r{/v3/netsim/([^/]+)/(\w+)/(\d+)$} do |shard_id, table_name, id|
+    dont_cache
+    content_type :json
+    TableType.new(netsim_channel_id, storage_id('shared'), channels_table_name(shard_id, table_name)).fetch(id.to_i).to_json
+  end
+
+  #
+  # DELETE /v3/netsim/<shard-id>/<table-name>/<row-id>
+  #
+  # Deletes a row by id.
+  #
+  delete %r{/v3/netsim/([^/]+)/(\w+)/(\d+)$} do |shard_id, table_name, id|
+    dont_cache
+    TableType.new(netsim_channel_id, storage_id('shared'), channels_table_name(shard_id, table_name)).delete(id.to_i)
+    no_content
+  end
+
+  #
+  # POST /v3/netsim/<shard-id>/<table-name>/<row-id>/delete
+  #
+  # This mapping exists for older browsers that don't support the DELETE verb.
+  #
+  post %r{/v3/netsim/([^/]+)/(\w+)/(\d+)/delete$} do |shard_id, table_name, id|
+    call(env.merge('REQUEST_METHOD'=>'DELETE', 'PATH_INFO'=>File.dirname(request.path_info)))
+  end
+
+  #
+  # POST /v3/netsim/<shard-id>/<table-name>
+  #
+  # Insert a new row.
+  #
+  post %r{/v3/netsim/([^/]+)/(\w+)$} do |shard_id, table_name|
+    unsupported_media_type unless request.content_type.to_s.split(';').first == 'application/json'
+    unsupported_media_type unless request.content_charset.to_s.downcase == 'utf-8'
+
+    value = TableType.new(netsim_channel_id, storage_id('shared'), channels_table_name(shard_id, table_name)).insert(JSON.parse(request.body.read), request.ip)
+
+    dont_cache
+    content_type :json
+
+    redirect "/v3/netsim/#{shard_id}/#{table_name}/#{value[:id]}", 301
+  end
+
+  #
+  # PATCH (PUT, POST) /v3/netsim/<shard-id>/<table-name>/<row-id>
+  #
+  # Update an existing row.
+  #
+  post %r{/v3/netsim/([^/]+)/(\w+)/(\d+)$} do |shard_id, table_name, id|
+    unsupported_media_type unless request.content_type.to_s.split(';').first == 'application/json'
```

Comment:
Since this is duplicated code with 78 and 79, please factor out the checks into a helper method so you can just say:

if not is_valid_media_type(request):
   unsupported_media_type

Or you could do it in the single line style used previously.  I prefer the two-line style in this case since unsupported_media_type is an exceptional case.